Riot/Utils/EventFormatter.m

```diff
@@ -408,6 +411,66 @@ - (NSString*)senderAvatarUrlForEvent:(MXEvent*)event withRoomState:(MXRoomState*
 
+#pragma mark - MXRoomSummaryUpdating
+- (BOOL)session:(MXSession *)session updateRoomSummary:(MXRoomSummary *)summary withStateEvents:(NSArray<MXEvent *> *)stateEvents roomState:(MXRoomState *)roomState
+{
+    BOOL updated = [super session:session updateRoomSummary:summary withStateEvents:stateEvents roomState:roomState];
+    
+    if (BuildSettings.supportFunctionalMembers)
+    {
+        // Customisation for EMS Functional Members
+        if ([self functionalMembersEventFromStateEvents:stateEvents])
+        {
+            MXLogDebug(@"[EventFormatter] The functional members event has been updated.")
+            return YES;
+        }
+    }
+    
+    return updated;
+}
+
+- (BOOL)session:(MXSession *)session updateRoomSummary:(MXRoomSummary *)summary withServerRoomSummary:(MXRoomSyncSummary *)serverRoomSummary roomState:(MXRoomState *)roomState
+{
+    BOOL updated = [super session:session updateRoomSummary:summary withServerRoomSummary:serverRoomSummary roomState:roomState];
+    
+    if (BuildSettings.supportFunctionalMembers)
+    {
+        // Customisation for EMS Functional Members
+        MXEvent *functionalMembersEvent = [self functionalMembersEventFromStateEvents:roomState.stateEvents];
+        
+        if (functionalMembersEvent)
+        {
+            MXLogDebug(@"[EventFormatter] Computing the room name and avatar excluding functional members.")
+            
+            NSArray<NSString*> *serviceMemberIDs = functionalMembersEvent.content[FunctionalMembersServiceMembersKey] ?: @[];
+            
+            updated |= [defaultRoomSummaryUpdater updateSummaryDisplayname:summary
+                                                                   session:session
+                                                     withServerRoomSummary:serverRoomSummary
+                                                                 roomState:roomState
+                                                          excludingUserIDs:serviceMemberIDs];
+            
+            updated |= [defaultRoomSummaryUpdater updateSummaryAvatar:summary
+                                                              session:session
+                                                withServerRoomSummary:serverRoomSummary
+                                                            roomState:roomState
+                                                     excludingUserIDs:serviceMemberIDs];
+        }
+    }
+
+    return updated;
+}
+
+/**
+ Gets the latest state event of type `io.element.functional_members` from the supplied array of state events.
+ @return An event of type `io.element.functional_members`, or nil if the event wasn't found.
+ */
+- (MXEvent *)functionalMembersEventFromStateEvents:(NSArray<MXEvent *> *)stateEvents
+{
+    NSPredicate *functionalMembersPredicate = [NSPredicate predicateWithFormat:@"type == %@", FunctionalMembersStateEventType];
+    return [stateEvents filteredArrayUsingPredicate:functionalMembersPredicate].lastObject;
```

Comment:
This operation will be expensive on big rooms like Matrix HQ. The request is for DM. Can we limit it to DM rooms?